### PR TITLE
Updated SoftmaxGrad and LogSoftmaxGrad to support version 13.

### PR DESCRIPTION
**Pull request by satyajandhyala**
**Description**: SoftmaxGrad needs to transpose inputs and output to implement version 13 as Softmax does.

**Motivation and Context**
- Why is this change required? What problem does it solve?
Currently only SoftmaxGrad and LogSoftmaxGrad exist upto version 12 although the forward operator Softmax and LogSoftmax support version 13. The gradient operators need to be modified or new gradients need to created to support training using version 13.
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cpu/op_gradients.cc [hunk 1]
```diff
@@ -74,80 +84,108 @@ Status SoftmaxGrad<T>::Compute(OpKernelContext* context) const {
...
-    math::Sum<float, CPUMathUtil>(d, dYdata + i * d, &sdY, nullptr, nullptr);
-    math::Axpy<float, CPUMathUtil>(d, -sdY, eYdata + i * d, dXdata + i * d, nullptr);
+  if (is_logsoftmaxgrad_) {
+    std::vector<float> eY(nd);
+    float* eYdata = eY.data();
+
+    // dX_ai = d(log Y_ai) - [sum_j d(log Y_aj)] exp(log Y_ai)
+    gsl::copy(gsl::make_span(dYdata, nd), gsl::make_span(dXdata, nd));
+    math::Exp<float, CPUMathUtil>(nd, Ydata, eYdata, nullptr);
+    for (size_t i = 0; i < N; ++i) {
```
Comment (xadupre) [inline orttraining/orttraining/training_ops/cpu/op_gradients.cc:144]: Why not parallelizing this loop too?

File: orttraining/orttraining/training_ops/cuda/math/softmax_grad.h [hunk 2]
```diff
@@ -15,15 +15,21 @@ template <typename T>
...
+        static_cast<const CUDAExecutionProvider*>(info.GetExecutionProvider()));
   }
 
   Status ComputeInternal(OpKernelContext* context) const override;
 
  private:
   int64_t axis_;
   bool log_softmax_;
+  int opset_;  // opset_ of the forward Softmax/LogSoftmax operator
+  CUDAExecutionProvider* cuda_ep_;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/math/softmax_grad.h:32]: what we actually need is just he cuda_ep->GetDeviceProp(). 
We can cache this, rather than the EP pointer. 
const cudaDeviceProp& prop

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 3]
```diff
@@ -1574,13 +1574,13 @@ TEST(GradientCheckerTest, SigmoidGrad) {
...
 
-void GradientCheckerSoftmaxGradHelper(bool is_log_softmax) {
+void GradientCheckerSoftmaxGradHelper(bool is_log_softmax, int version = 0) {
   TensorShape shape({3, 4, 5});
   float max_error;
   GradientChecker<float, float, float> gradient_checker;
 
   const std::string op = is_log_softmax ? "LogSoftmax" : "Softmax";
-  OpDef op_def{op};
+  OpDef op_def = version == 0 ? OpDef{op} : OpDef{op, kOnnxDomain, version};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:1583]: also need test case for axis = 1

Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:1583]: How about set the default value for version to 11? 

OpDef op_def = OpDef{op, kOnnxDomain, version};

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 4]
```diff
@@ -573,6 +573,51 @@ OpSchema& RegisterLambOpSchema(OpSchema&& op_schema) {
   return op_schema;
 }
 
+bool SoftMaxGradFunctionBodyBuilder(const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:576]: This function body was written for Softmax-11's gradient. 
Is the function body still valid for Softmax-11's gradient?

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc (hunk 3 above)
Comment (satyajandhyala) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:1583]: This is the existing UT. The default is set by the GradientChecker to 1 here
https://github.com/microsoft/onnxruntime/blob/b6fc9007f45f19119cd4bece7bce2a19d6d1f6a4/orttraining/orttraining/test/gradient/gradient_checker.cc#L300
Should we change it?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 5]
```diff
@@ -604,51 +649,27 @@ void RegisterTrainingOpSchemas() {
...
-          });
+  ONNX_CONTRIB_OPERATOR_SCHEMA(SoftmaxGrad_13)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "dY", "Gradient of output Y", "T")
+      .Input(1, "Y", "Input tensor", "T")
+      .Output(0, "dX", "Gradient of input X", "T")
+      .Attr(
+          "axis",
+          "Describes the axis of the inputs when coerced "
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:662]: The definition of axis should be updated to Softmax-13's.

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 6]
```diff
@@ -669,6 +690,25 @@ void RegisterTrainingOpSchemas() {
...
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(LogSoftmaxGrad_13)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "dY", "Gradient of output Y", "T")
+      .Input(1, "X", "Input tensor", "T")
+      .Output(0, "dX", "Gradient of input X", "T")
+      .Attr(
+          "axis",
+          "Describes the axis of the inputs when coerced "
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:701]: The definition of axis should be updated to LogSoftmax-13's.

File: orttraining/orttraining/training_ops/cpu/op_gradients.cc (hunk 1 above)
Comment (satyajandhyala) [reply orttraining/orttraining/training_ops/cpu/op_gradients.cc:144]: This is CPU gradient operation. This code exists only as a fallback.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 5 above)
Comment (satyajandhyala) [reply orttraining/orttraining/core/graph/training_op_defs.cc]: Fixed the description of axis attribute.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 6 above)
Comment (satyajandhyala) [reply orttraining/orttraining/core/graph/training_op_defs.cc]: Fixed the description of axis attribute.

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc (hunk 3 above)
Comment (satyajandhyala) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc]: Done

Comment (satyajandhyala) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc]: Added a new testcase for axis=1

File: orttraining/orttraining/training_ops/cuda/math/softmax_grad.h (hunk 2 above)
Comment (satyajandhyala) [reply orttraining/orttraining/training_ops/cuda/math/softmax_grad.h]: Done

File: orttraining/orttraining/training_ops/cuda/math/softmax_grad.h [hunk 7]
```diff
@@ -14,16 +14,21 @@ void dispatch_softmax_backward(cudaStream_t stream, output_t* grad_input, const
 template <typename T>
 class SoftmaxGrad final : public CudaKernel {
  public:
-  SoftmaxGrad(const OpKernelInfo& info) : CudaKernel{info} {
-    info.GetAttrOrDefault("axis", &axis_, static_cast<int64_t>(1));
-    log_softmax_ = info.GetKernelDef().OpName() == "LogSoftmaxGrad";
+  SoftmaxGrad(const OpKernelInfo& info) : CudaKernel{info},
+                                          prop_(const_cast<CUDAExecutionProvider*>(static_cast<const CUDAExecutionProvider*>(info.GetExecutionProvider()))->GetDeviceProp()) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/math/softmax_grad.h:18]: The const_cast might not be needed here?

**Review by SherlockNoMad — DISMISSED** [review]

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 8]
```diff
@@ -1574,23 +1574,29 @@ TEST(GradientCheckerTest, SigmoidGrad) {
...
 
-  // axis=0
+  // axis=1
   {
-    ASSERT_STATUS_OK(gradient_checker.ComputeGradientError(op_def, {shape}, {shape}, &max_error, {MakeAttribute("axis", int64_t(0))}));
+    ASSERT_STATUS_OK(gradient_checker.ComputeGradientError(op_def, {shape}, {shape}, &max_error, {MakeAttribute("axis", int64_t(1))}));
+    EXPECT_IS_TINY(max_error);
+  }
+
+  // axis=2
```
Comment (edgchen1) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:1597]: >// axis=2 [](http://example.com/codeflow?start=2&length=9)

is there a duplicate test for axis=2 below?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 9]
```diff
@@ -707,15 +707,15 @@ IMPLEMENT_GRADIENT_BUILDER(GetSigmoidGradient) {
 
 IMPLEMENT_GRADIENT_BUILDER(GetSoftmaxGradient) {
   return std::vector<NodeDef>{
-      NodeDef(OpDef{"SoftmaxGrad", kMSDomain, 1},
+      NodeDef(OpDef{SrcNodeOpsetVersion() < 13 ? "SoftmaxGrad" : "SoftmaxGrad_13", kMSDomain, 1},
```
Comment (edgchen1) [inline orttraining/orttraining/core/graph/gradient_builder.cc:710]: >"SoftmaxGrad_13" [](http://example.com/codeflow?start=65&length=16)

is versioning of gradient ops going to be handled by appending version numbers to the name?
why do we need the separate op? it's not obvious to me as the kernel registrations point to the same implementation, which also has logic for checking the opset version

Comment (satyajandhyala) [reply orttraining/orttraining/core/graph/gradient_builder.cc:710]: Yes. I understand that we want to keep the Contrib ops in MSDomain version #1 fixed. So are not versioning the contrib operators, including the gradient operators of ONNX operators. In the kernel implementation we ae checking the opset version of the forward operator.

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc (hunk 8 above)
Comment (satyajandhyala) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc]: Oops! I edited the wrong testcase after copy-paste.

**Review by edgchen1 — APPROVED** [review]
